src/light_data.rs

```diff
@@ -0,0 +1,307 @@
+        }
+    }
+}
+
+impl Arbitrary for LightData {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        let atom = prop::collection::vec(any::<u8>(), 0..256).prop_map(LightData::Atom);
+        atom.prop_recursive(16, 1024, 256, |inner| {
+            prop::collection::vec(inner.clone(), 0..256).prop_map(LightData::Cell)
+        })
+        .boxed()
+    }
+}
+
+impl LightData {
+    pub fn byte_count(x: usize) -> u8 {
+        if x == 0 {
+            return 1;
+        }
+        let mut n: u8 = 0;
+        let mut x = x;
+        while x > 0 {
+            n += 1;
+            x = x / 256;
+        }
+        n
+    }
+
+    pub fn to_trimmed_le_bytes(x: usize) -> Vec<u8> {
+        let mut res = vec![];
+        let mut x = x;
+        while x > 0 {
+            res.push((x % 256) as u8);
+            x = x / 256;
+        }
+        res
+    }
+    pub fn tag(&self) -> u8 {
+        match self {
+            Self::Atom(xs) if xs.len() == 0 => 0b0000_0000,
+            Self::Atom(xs) if xs.len() < 64 => 0b0100_0000u8 + (xs.len() as u8),
+            Self::Atom(xs) if xs.len() == 64 => 0b0100_0000u8,
+            Self::Atom(xs) => 0b0000_0000u8 + LightData::byte_count(xs.len()),
+            Self::Cell(xs) if xs.len() == 0 => 0b1000_0000,
+            Self::Cell(xs) if xs.len() < 64 => 0b1100_0000u8 + (xs.len() as u8),
+            Self::Cell(xs) if xs.len() == 64 => 0b1100_0000u8,
+            Self::Cell(xs) => 0b1000_0000u8 + LightData::byte_count(xs.len()),
+        }
+    }
+
+    pub fn tag_is_atom(x: u8) -> bool {
+        x & 0b1000_0000 == 0b0000_0000
+    }
+
+    pub fn tag_is_small(x: u8) -> bool {
+        x & 0b0100_0000 == 0b0100_0000
+    }
+
+    pub fn ser(&self) -> Vec<u8> {
+        let mut res = vec![];
+        res.push(self.tag());
+        match self {
+            Self::Atom(xs) if xs.len() == 0 => {}
+            Self::Atom(xs) if xs.len() <= 64 => res.extend(xs),
+            Self::Atom(xs) => {
+                res.extend(LightData::to_trimmed_le_bytes(xs.len()));
+                res.extend(xs);
+            }
+            Self::Cell(xs) if xs.len() == 0 => {}
+            Self::Cell(xs) if xs.len() <= 64 => {
+                for x in xs {
+                    res.extend(x.ser())
+                }
+            }
+            Self::Cell(xs) => {
+                res.extend(LightData::to_trimmed_le_bytes(xs.len()));
+                for x in xs {
+                    res.extend(x.ser())
+                }
+            }
+        };
+        res
+    }
+
+    pub fn de(i: &[u8]) -> Result<Self, ()> {
+        match Self::de_aux(i).finish() {
+            Ok((_, x)) => Ok(x),
+            Err(_) => Err(()),
+        }
+    }
+
+    pub fn de_aux(i: &[u8]) -> IResult<&[u8], Self> {
+        let (i, tag) = take(1u8)(i)?;
+        let tag = tag[0];
+        let size = tag & 0b11_1111;
+        if Self::tag_is_atom(tag) {
```

Comment:
Consider a `match (Self::tag_is_small(tag), Self::tag_is_atom(tag), size)` to simplify those cases.